kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/persistentOrderedMap/PersistentOrderedMapBuilder.kt

```diff
@@ -64,44 +64,43 @@ internal class PersistentOrderedMapBuilder<K, V>(private var map: PersistentOrde
     override fun put(key: K, value: @UnsafeVariance V): V? {
         val links = mapBuilder[key]
         if (links != null) {
-            if (links.value == value) {
+            if (links.value === value) {
                 return value
             }
-            mapBuilder[key] = LinkedValue(value, links.previous, links.next)
+            mapBuilder[key] = links.withValue(value)
             return links.value
         }
 
         if (isEmpty()) {  //  isEmpty
             firstKey = key
             lastKey = key
-            mapBuilder[key] = LinkedValue<K, V>(value, null, null)
+            mapBuilder[key] = LinkedValue(value)
             return null
         }
-        val oldLinks = mapBuilder[lastKey]!!
-        assert(oldLinks.next == null)
-        val newLinks = LinkedValue(oldLinks.value, oldLinks.previous, key)
+        val lastLink = mapBuilder[lastKey]!!
+        assert(lastLink.next == EndOfLink)
 
-        mapBuilder[lastKey as K] = newLinks
-        mapBuilder[key] = LinkedValue(value, lastKey, null)
+        mapBuilder[lastKey as K] = lastLink.withNext(key)
+        mapBuilder[key] = lastLink.putNextLink(value, lastKey)
         lastKey = key
         return null
     }
 
     override fun remove(key: K): V? {
         val links = mapBuilder.remove(key) ?: return null
 
-        if (links.previous != null) {
+        if (links.previous !== EndOfLink) {
             val previousLinks = mapBuilder[links.previous]!!
-            assert(previousLinks.next == key)
-            mapBuilder[links.previous] = LinkedValue(previousLinks.value, previousLinks.previous, links.next)
+//            assert(previousLinks.next == key)
+            mapBuilder[links.previous as K] = previousLinks.withNext(links.next)
         }
-        if (links.next != null) {
+        if (links.next !== EndOfLink) {
             val nextLinks = mapBuilder[links.next]!!
-            assert(nextLinks.previous == key)
-            mapBuilder[links.next] = LinkedValue(nextLinks.value, links.previous, nextLinks.next)
+//            assert(nextLinks.previous == key)
+            mapBuilder[links.next as K] = nextLinks.withPrevious(links.previous)
         }
-        firstKey = if (key == firstKey) links.next else firstKey
-        lastKey = if (key == lastKey) links.previous else lastKey
+        firstKey = if (links.previous === EndOfLink) links.next as? K else firstKey
+        lastKey = if (links.next === EndOfLink) links.previous as? K else lastKey
```

Comment:
Makes sense to move these to corresponding `else` blocks, to avoid `firstKey = firstKey` assignment.

Reply:
+1